crates/astria-sequencer-relayer/src/relayer/write/mod.rs

```diff
@@ -426,14 +503,94 @@ async fn submit_with_retry(
 
     let blobs = Arc::new(blobs);
 
-    let height = tryhard::retry_fn(move || {
-        client
-            .clone()
-            .try_submit(blobs.clone(), last_error_receiver.clone())
+    let final_state = tryhard::retry_fn(move || {
+        try_submit(
+            client.clone(),
+            blobs.clone(),
+            started_submission.clone(),
+            largest_sequencer_height,
+            last_error_receiver.clone(),
+        )
     })
     .with_config(retry_config)
     .in_current_span()
     .await
-    .wrap_err("retry attempts exhausted; bailing")?;
-    Ok(height)
+    .wrap_err("finished trying to submit")?;
+    Ok(final_state)
+}
+
+#[instrument(skip_all)]
+async fn try_submit(
+    mut client: CelestiaClient,
+    blobs: Arc<Vec<Blob>>,
+    started_submission: StartedSubmission,
+    largest_sequencer_height: SequencerHeight,
+    last_error_receiver: watch::Receiver<Option<SubmissionError>>,
+) -> Result<StartedSubmission, SubmissionError> {
+    // Get the error from the last attempt to `try_submit`.
+    let maybe_last_error = last_error_receiver.borrow().clone();
+    let maybe_try_submit_error = match maybe_last_error {
+        // If error is broadcast timeout, try to confirm submission from last attempt.
+        Some(SubmissionError::BroadcastTxTimedOut(prepared_submission)) => {
+            if let Some(new_state) =
+                try_confirm_submission_from_failed_attempt(client.clone(), prepared_submission)
+                    .await?
```

Comment:
context

Reply:
This is not an `eyre` error type (we need to match on the variant so I used an enum).

Reply:
Gotcha